lib/server.js

```diff
@@ -386,6 +406,88 @@ module.exports = function createServer(config, log) {
     }
   )
 
+  api.post('/checkCustomRule', function(req, res, next) {
+    const action = req.body.action
+    const ip = req.body.ip
+    const email = req.body.email
+
+    if (!ip || !action || !email) {
+      const err = {code: 'MissingParameters', message: 'action, ip email, are all required'}
+      log.error({op: 'request.checkCustomRule', action, ip, email})
+      res.send(400, err)
+      return next()
+    }
+
+    // Get all the custom rules that might apply to this action
+    const rules = config.userDefinedRateLimitRules
+    const checkRules = computedRules.get(action)
+
+    if (checkRules.length <= 0) {
+      const err = {code: 'MissingParameters', message: 'no rules found for ' + action}
+      log.error({op: 'request.checkCustomRule', action, ip, email})
+      res.send(400, err)
+      return next()
+    }
+
+    // Retrieve the record for each rule and check if `action` has been rate limited
+    const recordResults = []
+    return P.each(checkRules, (ruleName) => {
+      let retryAfter = null
+      const recordKey = ip + utils.createHashHex(email, ruleName)
+      return mc.getAsync(recordKey)
+        .then((object) => {
+          return new Record(object, rules[ruleName])
+        }, () => {
+          return new Record({}, rules[ruleName])
+        })
+        .then((record) => {
+          retryAfter = record.update(action)
+          const minLifetimeMS = record.getMinLifetimeMS()
+
+          // To save space in memcache, don't store limits and
+          // actions since they can be referenced from config
+          record.limits = null
+          record.actions = null
+
+          const lifetime = Math.max(config.memcache.recordLifetimeSeconds, minLifetimeMS / 1000)
+          return mc.setAsync(recordKey, record, lifetime)
+        })
+        .then(() => {
+          recordResults.push({
+            block: retryAfter > 0,
```

Comment:
Do you actually use this `block` key in the logic below?  I wonder if this would be simpler as just a list of `retryAfter` integers, rather than a list of rich objects with a `retryAfter` key.